### PR TITLE
[Draft] KAFKA-10532: close clean for EOS when it's RUNNING standby or RESTORING active

### PR DESCRIPTION
**Pull request by lqxshay**
Fixes https://issues.apache.org/jira/browse/KAFKA-10532

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rishiraj88** [comment]
@lqxshay , thanks for for the helpful checklist.

**Review by guozhangwang — COMMENTED** [review]
Hello @lqxshay , thanks for the PR.

I took a quick look at it and here are some thoughts: when closing a task-dirty, it means the task itself may have inconsistent states and hence under EOS, it may violates the guarantees. In that case we'd have to be conservative and wipe out the stores.

For tasks that are not actively processing though, i.e. if it is an active task still being restored (in the new model, it means it is still inside the StateUpdater), or a standby task being updated, then we do not need to play conservatively since we know the state would still be consistent anyways.

So I think the check should be around if the task is being actively processed (i.e. inside TaskManager), or being restored (i.e. inside StateUpdater).

More specifically, the code related to it would be in `TaskManager#handlRemovedTasksFromStateUpdater` since those tasks from the state updater, can ignore the EOS flag even when they are about to close dirty.

**Comment by lqxshay** [comment]
@guozhangwang Apologies for the delayed response here. 

Thanks for the explanation. To put it in another way, as long as the task is in certain states(RESTORING active, RUNNING standby task), we would always close clean. I've updated the PR to disregard PendingUpdateAction.Action.CLOSE_DIRTY for the condition in `TaskManager#handlRemovedTasksFromStateUpdater` and `TaskManager#handleRestoredTasksFromStateUpdater`.

Could you please take another look?

**Review by guozhangwang — COMMENTED** [review]
Thanks @lqxshay .  I think the PR is going towards the right directions. A couple more thoughts:

1. I'd admit there are a lack of detailed state transition diagram as for StreamTask and StandbyTask besides the on we had in the Task interface itself. In general, I think you can take that digram and keep in mind that for standby task, it would not have a `RESTORING` state, but just `RUNNING`. To give you a bit more information: when the active task (StreamTask) and standby task are inside the state updater, their states should be `RESTORING` and `RUNNING`. That's because the state transition to `RESTORING` and `RUNNING` respectively happens in `initializeIfNeeded` function, and that is called right before the task is given to the state updater. Hence, when we get them back from the state updater because these tasks need to be closed (dirty), they should always be in that state and we do not need to do the check in `handleRemovedTasksFromStateUpdater`.
2. We'd need to do the check while we call `closeTaskDirty` on the collected task one-by-one though --- note that here we still need to call the `closeTaskDirty` not the `closeTaskClean`, but all we need is to avoid the wiping state. They tricky thing is that inside that function we first call `suspend` to transit the task state to `SUSPENDED` and then call `closeDirty`, so when we are inside `closeDirty` the state has already been in `SUSPENDED` and hence we cannot tell if it was previously `RESTORING` or not. 

So here's my thoughts:
1. We combine the two boolean `eosEnabled` and `closeClean` of the util func `StateManagerUtil.closeStateManager` into a single flag, since they are only used to be consolidated into a `final boolean wipeStateStore = !closeClean && eosEnabled`. So why not just pass in the `wipeStateStore` directly into it.
2. For StandbyTask, it is easy: we can always set `wipeStateStore` to false when calling the util func inside `closeDirty` since it does not have a `RUNNING` state at all.
3. For StreamTask, we need to put some effort: we can add a flag like `suspendedFromRestoring` and set it in the `suspend` func if the prev state is `RESTORING`, and reset it if we do transit from `SUSPENDED` to any other state rather than calling `close`. then in `closeDirty` we set the `wipeStateStore = !suspendedFromRestoring && !closeClean && eosEnabled`. I know this is a bit hacky but given that in the new rebalance protocol the whole `SUSPEND` state will be gone and we can transit from `RESTORING` to `CLOSED` directly this flag would be removed as well.

Also a meta comment: we would need to add some unit test coverage for this new behavior when the PR is finally open for merge :) also cc @mjsax @lucasbru who's familiar with EOS and want to keep them in the loop.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java [hunk 1]
```diff
@@ -276,6 +276,11 @@ private void close(final boolean clean) {
         transitionTo(State.CLOSED);
     }
 
+    @Override
+    boolean shouldWipeStateStore(final boolean closeClean, final boolean eosEnabled, final boolean suspendedFromRestoring) {
+        return closeClean;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:281]: Hmm.. not sure if this is correct, let me think through this.

**Review by guozhangwang — COMMENTED** [review]
I looked at the statea-updater for various exceptions thrown:

1) task-corrupted exception: which is translated from invalid-offset exception from restore consumer's poll)
2) streams exception: translated from kafka exception from restore consumer's other calls, as well as the restore callback)
3) runtime exception: any other exceptions.

I'm wondering now if we can safely skip wiping out state stores for all such cases which will cause the task to be closed dirty. For 2) I think we can skip the wiping for sure; for 1) it may indicate the local restored state may have been "polluted" with obsoleted record, and hence we'd have to wipe; for 3) it is hard to tell, e.g. if it's from an IO exception then maybe we do not need to wipe.

With all that in mind, I think until we can fine categorize all the exceptions that cause the task to be closed dirty, we cannot just always do this optimization? WDYT @mjsax @cadonna

**Comment by lucasbru** [comment]
> 2. streams exception: translated from kafka exception from restore consumer's other calls, as well as the restore callback)

`InvalidOffsetException` is rethrown as `StreamsException` in `prepareChangelogs` and `hasRestoredToEnd`. That seems to violate your categorization.

**Comment by mjsax** [comment]
Closing this PR as slate. The Jira ticket was reassigned to @eduwercamacaro who is picking up this work.

**Pull request closed without merging**